lib/uncruft/deprecate_attribute.rb

```diff
@@ -0,0 +1,30 @@
+module Uncruft
+  module DeprecateAttribute
+    extend ActiveSupport::Concern
+
+    module ClassMethods
+      def deprecate_attribute(attribute, aliased_attribute: nil, message:)
```

Comment:
Would be interested to hear feedback on `aliased_attribute` naming from someone with fresh 👀

Reply:
I'm wondering whether this aliasing feature will cause confusion, and if there's anything we can do to anticipate that.

Basically, I think there are two ways of reading `aliased_attribute`:

1. "Please alias this for me now" (the current implementation)
2. "I have already implemented the other/aliased method but I want to indicate that this method is deprecated in favor of that other one."

In the case of number 2, it would result in someone potentially overwriting a method that they already defined, and unexpected behavior in general.

I also left [a note about potentially auto-generating the message](https://github.com/Betterment/uncruft/pull/4/files#r461027022), but I think that opens up a can of worms when it comes to the "use X insetad" case, because the ability to specify an alternate, non-deprecated method (without having this framework auto-alias it) feels pretty reasonable in that case.

Reply:
👍 that's a very valid concern. The tense there is maybe part of the issue. Uncruft doesn't want you to provide the name of an alias you created - it wants to create the alias for you.

I also wonder if we should be less focused on what we're going to do internally with the argument you provide (create an alias of the deprecated method by this name) and more focused on how it will alter the interface of the class (add a "safe", deprecation-less version of the deprecated method by this name).

What about this?

```ruby
def deprecate_attribute(attribute, safe_name: nil, message:)
```

Reply:
`safe_name` is even less clear to me though, because I still read it as "deprecate 'X', because the safe name (that I have already implemented) is 'Y'"

What if we turn the intent entirely on its head and make the API look like this:

```ruby
alias_attribute :nickname, :name, deprecation: "`name` is deprecated. Please use `nickname` instead"
```

Reply:
(e.g. extend the existing alias method with a new deprecation arg)

Reply:
Oh, interesting. I like that. We'd be giving up a couple things, though:

- Deprecating non-attribute methods
- DSL for deprecating non-aliased attributes

I'm more concerned about the first one, as I can see aliased deprecation use cases for non-ActiveModel-attribute methods (e.g. updated balance calculation, Her model attributes, etc).

Less concerned about the second case, but I do think there's something about a DSL that would drive adoption for using deprecations internally in a way that dropping down one level to `ActiveSupport::Deprecation.warn` doesn't.

Reply:
I'd propose combining those two bullets into a single "deprecate this method" helper, and then wrapping that with my proposed `alias_attribute` change -- basically the goal is so that the "deprecate this method" helper never _also_ aliases, but the "alias this method" helper can _also_ optionally deprecate the old one.

Main goal there is to try to solve the "what should do the aliasing" problem intuitively and sidestep the problem of how to frame/describe/name the `aliased_attribute` argument since it would no longer exist.

Reply:
Love it!

So, we would have:

```ruby
deprecate_method :name, message: '`name` is deprecated. Users are now anonymous.'
```

and

```ruby
alias_attribute :nickname, :name, deprecation: "`name` is deprecated. Please use `nickname` or `formal_name` instead."
```

That sounds great to me. I could still maybe see having a `deprecate_attribute` method without the alias, just so we have a shorthand for:

```ruby
deprecate_method :name, message: '`name` is deprecated. Users are now anonymous.'
deprecate_method :name=, message: '`name` is deprecated. Users are now anonymous.'
```

by writing:

```ruby
deprecate_attribute :name, message: '`name` is deprecated. Users are now anonymous.'
```

Side note: If we do eventually decide we're comfortable with default deprecation messages, it would be pretty cool that this would work with the same ergonomics as `Memoist`:

```ruby
deprecate_method def name
  [first_name, middle_initial, last_name].compact.join(' ')
end
```

Reply:
Worth noting -- `alias_attribute` also aliases the `?` methods, so `deprecate_attribute` could do the same!!!